pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,353 @@
+	// other workers.
+	defer c.resourceQueue.Done(key)
+
+	if err := c.processResource(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.resourceQueue.AddRateLimited(key)
+		return true
+	}
+	c.resourceQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) processNextNamespace(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.namespaceQueue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.namespaceQueue.Done(key)
+
+	if err := c.processNamespace(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.namespaceQueue.AddRateLimited(key)
+		return true
+	}
+	c.namespaceQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) processNextCluster(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.clusterQueue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.clusterQueue.Done(key)
+
+	if err := c.processCluster(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.clusterQueue.AddRateLimited(key)
+		return true
+	}
+	c.clusterQueue.Forget(key)
+	return true
+}
+
+// key is gvr::KEY
+func (c *Controller) processResource(ctx context.Context, key string) error {
+	parts := strings.SplitN(key, "::", 2)
+	if len(parts) != 2 {
+		klog.Errorf("Error parsing key %q; dropping", key)
+		return nil
+	}
+	gvrstr := parts[0]
+	gvr, _ := schema.ParseResourceArg(gvrstr)
+	if gvr == nil {
+		klog.Errorf("Error parsing GVR %q; dropping", gvrstr)
+		return nil
+	}
+	key = parts[1]
+
+	obj, exists, err := c.ddsif.IndexerFor(*gvr).GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		klog.Infof("object %q does not exist", key)
+		return nil
+	}
+	unstr, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.Errorf("object was not Unstructured, dropping: %T", obj)
+		return nil
+	}
+	unstr = unstr.DeepCopy()
+
+	// Get logical cluster name.
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("failed to split key %q, dropping: %v", key, err)
+		return nil
+	}
+	lclusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+	return c.reconcileResource(ctx, lclusterName, unstr, gvr)
+}
+
+// namespaceBlocklist holds a set of namespaces that should never be synced from kcp to physical clusters.
+var namespaceBlocklist = sets.NewString("kube-system", "kube-public", "kube-node-lease")
+
+func (c *Controller) processNamespace(ctx context.Context, key string) error {
+	ns, err := c.namespaceLister.Get(key)
+	if err != nil {
+		return err
```

Comment:
If not found error, return nil (namespace was deleted)

Reply:
Will that be a problem if the lister's cache is just out of date? It might return not-found when it actually does exist, and we'll drop this update.

Reply:
The lister's cache is guaranteed to be at least as fresh as the event that triggered the shared informer's add/update/delete event handler.

This is more of a problem if you're doing a write operation and then you quickly/immediately expect that a lister's list/get invocation would show the changes from your write - those can be delayed.

Does this help?

Reply:
Very helpful! Thanks, dropping `IsNotFound` now. 👍